### ISSUE TITLE
[Feature Request] Better KL Divergence Approximation

### ISSUE THREAD

**Issue by 09tangriro** [DLR-RM/stable-baselines3#417]
**Important Note: We do not do technical support, nor consulting** and don't answer personal questions per email.
Please post your question on the [RL Discord](https://discord.com/invite/xhfNqQv), [Reddit](https://www.reddit.com/r/reinforcementlearning/) or [Stack Overflow](https://stackoverflow.com/) in that case.

### 🚀 Feature

Use a better estimate for the KL Divergence in the PPO algorithm. 

The estimator I propose is the reverse estimator here:  
![image](https://user-images.githubusercontent.com/47002898/116915503-04032280-ac44-11eb-9448-5f015ed88885.png)

This has a lower variance since the extra term is negatively correlated and also is positive semi-definite (see motivation section).

### Motivation

In the PPO algorithm, there is a KL limit method used as a final block to large updates to the policy in a single timestep. The line used is: `approx_kl_divs.append(th.mean(rollout_data.old_log_prob - log_prob).detach().cpu().numpy())`. This is an unbiased estimator, but it has large variance since it can take on negative values (as opposed to the actual KL Divergence measure)! This can cause problems in the check: `if self.target_kl is not None and np.mean(approx_kl_divs) > 1.5 * self.target_kl:` since this doesn't consider the large negative values that `approx_kl_divs` could take. 

### Pitch

I want to replace the KL Divergence equation currently used in the PPO algorithm with the better approximation described above.

NOTE: This is basically a 1 line change.

### Alternatives

N/A

### Additional context

![image](https://user-images.githubusercontent.com/47002898/116917113-2433e100-ac46-11eb-869a-fe984ad164ac.png)

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
Could you provide link to source on that text (mostly for my interest)? :). This is a one line change but will change results of codes for people who run things, so need to think twice before adding it.

**Comment by 09tangriro** [comment]
Sure! Here's a blog post on the topic by John Schulman: http://joschu.net/blog/kl-approx.html

**Comment by Miffyli** [comment]
Thanks for the link! Was very informative read :). 

I say we update the code with this version to encourage better solutions. Leave links to these sources (this issue and Schulman's blog post) right beside the KL computation.

**Comment by 09tangriro** [comment]
Great! I'll get started on a PR :)

**Comment by araffin** [comment]
> I say this is a breaking-change level of change, because this will change the training behaviour for anybody who uses target_kl argument.

yes, btw, could you document in the PR on a small example (for instance cartpole) how different are the two estimators in practice?

I'm also not 100% sure that the way I wrote the check for the target kl is right ^^#... (current checked after every epoch and not before each gradient update)

**Comment by araffin** [comment]
> since this doesn't consider the large negative values that approx_kl_divs could take.

just one question: kl div cannot be negative, right? or is it only the approximation?

**Comment by Miffyli** [comment]
> just one question: kl div cannot be negative, right? or is it only the approximation?

Right, it is only the (naive) approximation currently used that can lead to negative values.

**Comment by 09tangriro** [comment]
Sure I can run a little experiment :)

Indeed, the actual KL divergence cannot be negative, but the approximation used here can be negative. Therefore, when checking if the KL divergence is too large, I think it should check the absolute value of the approximation with respect to the positive `target_kl,` (do correct me if I've missed something!). This won't be an issue with the new estimator though which also cannot take negative values.

Also, I think we should use the reverse estimate (KL Divergence is not an even function!) which is mode seeking rather than mean seeking :)

**Comment by 09tangriro** [comment]
See this article for some good stuff on forward vs reverse KL: https://dibyaghosh.com/blog/probability/kldivergence.html

**Comment by Miffyli** [comment]
> I'm also not 100% sure that the way I wrote the check for the target kl is right ^^#... (current checked after every epoch and not before each gradient update)

For better constraints, it could be done before every gradient step, but I wonder if this becomes too noisy (i.e. easily terminates prematurely if `target_kl` is reached just by random sampling). Then again, this is probably the most intuitive approach, so I say we update the `target_kl` check to be before every gradient step :D

**Comment by araffin** [comment]
>For better constraints, it could be done before every gradient step, but I wonder if this becomes too noisy

yes, that was the main reason I chose to wait after each epoch.

looking at John Schulman code, it is even done differently: https://github.com/joschu/modular_rl/blob/master/modular_rl/ppo.py#L48

**Comment by 09tangriro** [comment]
One last thing, instead of having a `target_kl`, an idea might be to have `max_kl` so that we don't need to calculate `1.5 * self.target_kl` each time when checking `if self.target_kl is not None and np.mean(approx_kl_divs) > 1.5 * self.target_kl:`. In this case, the user would define that max instead of the target, what do you think?

**Comment by Miffyli** [comment]
`max_kl` does sound more reasonable (that is what trust-region PGs do), but I would focus on updating KL computation for now and leave the parameter (and its behaviour) as is, because that would be a breaking backwards-compat behaviour.
